packages/vite/src/node/ssr/ssrModuleLoader.ts

```diff
@@ -174,18 +182,36 @@ async function instantiateModule(
   return Object.freeze(ssrModule)
 }
 
-function nodeRequire(
+// In node@12+ we can use dynamic import to load CJS and ESM
+// Caveat https://github.com/vitejs/vite/pull/5197#discussion_r722583544
+async function nodeImport(
   id: string,
   importer: string | null,
-  root: string,
-  preserveSymlinks: boolean
+  config: ViteDevServer['config'],
 ) {
-  const mod = require(resolve(id, importer, root, preserveSymlinks))
-  const defaultExport = mod.__esModule ? mod.default : mod
-  // rollup-style default import interop for cjs
+  const url = pathToFileURL(resolve(id, importer, config.root, !!config.resolve.preserveSymlinks))
+  const mod = await nativeImport(url);
+  return proxyESM(mod);
+}
+
+async function handleImport(
+  id: string,
+  importer: string | null,
+  config: ViteDevServer['config'],
+): Promise<any> {
+    // `resolve` doesn't handle `node:` builtins, so handle them directly
+    if (id.startsWith('node:') || builtins.has(id)) {
+      return nativeImport(id);
+    }
+    return nodeImport(id, importer, config);
+}
+
+// rollup-style default import interop for cjs
+function proxyESM(mod: any) {
+  const isESM = !!(mod.__esModule ?? mod.default);
   return new Proxy(mod, {
     get(mod, prop) {
-      if (prop === 'default') return defaultExport
+      if (prop === 'default') return isESM ? mod.default : mod;
```

Comment:
Just a callout: this is also not spec compliant but likely matches user expectations, thus the "rollup-style" comment.